Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -1162,9 +1172,94 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+    m_selectionTrackingData.shouldExtendCurrentSelection = mouseEventModifiers.contains(WebEventModifier::ShiftKey);
+
+    if (m_selectionTrackingData.shouldMakeMarqueeSelection)
+        return continueTrackingSelection(pageIndex, pagePoint);
+
+    // FIXME: <https://webkit.org/b/268617>  Selections should be extensible on shift-click.
+    if (m_selectionTrackingData.shouldExtendCurrentSelection)
+        notImplemented();
+
+    continueTrackingSelection(pageIndex, pagePoint);
+}
+
+static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, const WebCore::IntPoint& point2)
+{
+    auto marqueeRectLocation { point1.shrunkTo(point2) };
+    IntSize marqueeRectSize { point1 - point2 };
+    return { marqueeRectLocation.x(), marqueeRectLocation.y(), std::abs(marqueeRectSize.width()), std::abs(marqueeRectSize.height()) };
+}
+
+void UnifiedPDFPlugin::continueTrackingSelection(PDFDocumentLayout::PageIndex pageIndex, const WebCore::IntPoint& pagePoint)
+{
+    m_selectionTrackingData.isActive = true;
+
+    if (m_selectionTrackingData.shouldMakeMarqueeSelection) {
+        if (m_selectionTrackingData.startPageIndex != pageIndex)
+            return;
+
+        m_selectionTrackingData.marqueeSelectionRect = computeMarqueeSelectionRect(pagePoint, m_selectionTrackingData.startPagePoint);
+        auto page = m_documentLayout.pageAtIndex(pageIndex);
+        return setCurrentSelection([page.get() selectionForRect:m_selectionTrackingData.marqueeSelectionRect]);
+    }
+
+    switch (m_selectionTrackingData.granularity) {
+    case SelectionGranularity::Character: {
+        auto fromPage = m_documentLayout.pageAtIndex(m_selectionTrackingData.startPageIndex);
+        auto toPage = m_documentLayout.pageAtIndex(pageIndex);
+        RetainPtr<PDFSelection> selection = [m_pdfDocument selectionFromPage:fromPage.get() atPoint:m_selectionTrackingData.startPagePoint toPage:toPage.get() atPoint:pagePoint];
+        if (m_selectionTrackingData.shouldExtendCurrentSelection)
+            [selection addSelection:m_selectionTrackingData.selectionToExtendWith.get()];
+        setCurrentSelection(WTFMove(selection));
+        break;
+    }
+    // FIXME: <https://webkit.org/b/268616> Selection tracking should be able to reason at word/line granularity.
+    case SelectionGranularity::Word:
+    case SelectionGranularity::Line:
+        notImplemented();
+    }
+}
+
+void UnifiedPDFPlugin::setCurrentSelection(RetainPtr<PDFSelection>&& selection)
+{
+    m_currentSelection = selection;
+    if (!m_selectionTrackingData.isActive)
+        commitCurrentSelection(CommitReason::SelectionIsNoLongerActive);
+}
+
+void UnifiedPDFPlugin::commitCurrentSelection(CommitReason reason)
+{
+    if (reason == CommitReason::ReceivedMouseUp && !std::exchange(m_selectionTrackingData.isActive, false))
+        return;
+    LOG(PDF, "Committing text selection: %s", m_currentSelection.get().string.UTF8String);
+    notifySelectionChanged();
+    updateLayerHierarchy();
```

Comment:
This is unnecessary. I'll remove it.

Reply:
Fixed in [53cae5f](https://github.com/WebKit/WebKit/commit/53cae5fe11eca85d9ea822cdec0181e4a4f4dcf6)